src/qibo/validation/gate_set_tomography.py

```diff
@@ -0,0 +1,2089 @@
+        nqubits = 1
+        idx_basis = ["I", "X", "Y", "Z"]
+
+        result = circuit.execute(nshots=NshotsGST)
+        counts = dict(result.frequencies(binary=True))
+
+        if save_data is not None:
+            import os
+
+            filedirectory = "GST_basisops_qibo_1qb/"
+            fname = "BasisOp%.2d (k=%.2d,j=%.2d) %s_basis.txt" % (
+                idx_basis_op,
+                k,
+                j,
+                idx_basis[j],
+            )
+            # If location does not exist, create location.
+            GST_basisops_qibo_1qb = "GST_basisops_qibo_1qb"
+            if not os.path.exists(GST_basisops_qibo_1qb):
+                os.makedirs(GST_basisops_qibo_1qb)
+
+            # Save counts into a matrix. 0th column: word; 1st column: count value
+            # --------------------------------------------------------------------
+            fout = filedirectory + fname
+            fo = open(fout, "w")
+            for v1, v2 in counts.items():
+                fo.write(str(v1) + "  " + str(v2) + "\n")
+            fo.close()
+
+        # Find value for Bjk tilde
+        # -------------------------
+
+        expectation_signs = np.matrix([[1, 1, 1, 1], [1, -1, -1, -1]])
+
+        matrix = np.zeros((2**nqubits, 2**nqubits))
+        for ii in range(0, 2):
+            word = bin(ii)[2:]
+            matrix[ii, 0] = word
+
+        for v1, v2 in counts.items():
+            # print(v1, v2)
+            row = int(v1, 2)
+            col = v2
+            matrix[row, 1] = col
+
+        sorted_matrix = sort_counts_1qb(matrix)
+        probs = sorted_matrix[:, 1] / np.sum(sorted_matrix[:, 1])
+        sorted_matrix = np.column_stack((sorted_matrix, probs))
+
+        k = k + 1
+        j = j + 1
+
+        if j == 1:
+            sorted_matrix = np.column_stack((sorted_matrix, expectation_signs[:, 0]))
+        elif j == 2:
+            sorted_matrix = np.column_stack((sorted_matrix, expectation_signs[:, 1]))
+        elif j == 3:
+            sorted_matrix = np.column_stack((sorted_matrix, expectation_signs[:, 1]))
+        elif j == 4:
+            sorted_matrix = np.column_stack((sorted_matrix, expectation_signs[:, 1]))
+        elif j == 5:
+            sorted_matrix = np.column_stack((sorted_matrix, expectation_signs[:, 2]))
+        elif j == 9:
+            sorted_matrix = np.column_stack((sorted_matrix, expectation_signs[:, 2]))
+        elif j == 13:
+            sorted_matrix = np.column_stack((sorted_matrix, expectation_signs[:, 2]))
+        else:
+            sorted_matrix = np.column_stack((sorted_matrix, expectation_signs[:, 3]))
+
+        temp = np.array(sorted_matrix[:, 2]) * np.array(sorted_matrix[:, 3])
+        sorted_matrix = np.column_stack((sorted_matrix, temp))
+
+        expectation_val = np.sum(temp)
+
+        return expectation_val
+
+    def GSTBasisOpMeasurements_13(circ):
+        nqubits = 1
+        initial_state = np.zeros(2**nqubits)
```

Comment:
Better to avoid using global variables inside of functions (``nqubits``).

Reply:
The ```GSTBasisOpMeasurements_13``` function is written for a single qubit, while ```GSTBasisOpMeasurements_241``` is written for two qubits. They can't be used interchangeably, meaning that if I have a single qubit system, I cannot call ```GSTBasisOpMeasurements_241```, vice versa.

Hence I thought to use ```nqubits = 1``` or ```nqubits = 2``` within these separate functions. ```nqubits``` will not be global variables since they're within the functions, will they?

Reply:
No, sorry, my bad I thought you were using the nqubits variable defined above. Still you have several redefinitions of ``nqubits`` throughout the code which are usually useless, since they are fixed in the chunck of code where they are used, why not to use the value directly where it's needed then? In general as I said above I would avoid the definition of functions inside of other functions unless their scope is very minimal. So take this out as well please.

Reply:
Got it, I will define ```nqubits =``` whenever needed only.